compiler/src/dotty/tools/dotc/transform/localopt/RemoveUnnecessaryNullChecks.scala

```diff
@@ -6,104 +6,120 @@ import core.Contexts.Context
 import core.Symbols._
 import core.Types._
 import core.Flags._
+import core.Names._
+import core.StdNames._
 import ast.Trees._
 import scala.collection.mutable
 
-/** Eliminated null checks based on the following observations:
+/** Eliminate unnecessary null checks
  *
  *  - (this)  cannot be null
  *  - (new C) cannot be null
  *  - literal is either null itself or non null
  *  - fallsback to `tpe.isNotNull`, which will eventually be true for non nullable types.
  *  - in (a.call; a == null), the first call throws a NPE if a is null; the test can be removed.
  *
- *  @author DarkDimius, Jvican, OlivierBlanvillain
+ *  
+ *  @author DarkDimius, Jvican, OlivierBlanvillain, gan74
  */
- class RemoveUnnecessaryNullChecks extends Optimisation {
+class RemoveUnnecessaryNullChecks extends Optimisation {
   import ast.tpd._
 
-  val initializedVals = mutable.HashSet[Symbol]()
-
-  val checkGood = newMutableSymbolMap[Set[Symbol]]
-
-  def clear(): Unit = {
-    initializedVals.clear()
-    checkGood.clear()
-  }
-
-  def isGood(t: Symbol)(implicit ctx: Context): Boolean = {
-    t.exists && initializedVals.contains(t) && {
-      var changed = true
-      var set = Set(t)
-      while (changed) {
-        val oldSet = set
-        set = set ++ set.flatMap(x => checkGood.getOrElse(x, Nil))
-        changed = set != oldSet
+  // matches sym == null or null == sym
+  object NullCheck {
+    def unapply(t: Apply)(implicit ctx: Context): Option[Symbol] =
+      t match {
+        case t @ Apply(Select(id: Ident, op), List(rhs)) if (t.symbol == defn.Object_eq || t.symbol == defn.Any_==) && !isVar(id.symbol) && isAlwaysNull(rhs) => Some(id.symbol)
+        case t @ Apply(Select(lhs, op), List(id: Ident)) if (t.symbol == defn.Object_eq || t.symbol == defn.Any_==) && !isVar(id.symbol) && isAlwaysNull(lhs) => Some(id.symbol)
+        case _ => None
       }
-      !set.exists(x => !initializedVals.contains(x))
-    }
   }
 
-  def visitor(implicit ctx: Context): Tree => Unit = {
-    case vd: ValDef =>
-      val rhs = vd.rhs
-      if (!vd.symbol.is(Mutable) && !rhs.isEmpty) {
-        def checkNonNull(t: Tree, target: Symbol): Boolean = t match {
-          case Block(_ , expr) =>
-            checkNonNull(expr, target)
+  def clear(): Unit = ()
 
-          case If(_, thenp, elsep) =>
-            checkNonNull(thenp, target) && checkNonNull(elsep, target)
+  def visitor(implicit ctx: Context): Tree => Unit = NoVisitor
 
-          case _: New | _: This => true
 
-          case t: Apply if t.symbol.isPrimaryConstructor => true
-
-          case t: Literal => t.const.value != null
+  def transformer(implicit ctx: Context): Tree => Tree = {
+    // transform tree recursively replacing nullchecks for symbols in nullness
+    def transform(tree: Tree, nullness: mutable.Map[Symbol, Boolean]) =
+      if (nullness.isEmpty) tree
+      else new TreeMap() {
+          override def transform(tree: Tree)(implicit ctx: Context): Tree = {
```

Comment:
You are right. However nullness isn't static, even for `val`, it can change within a block. There is no way to evaluate if a value is null without being aware of what code has been executed before. 

I wasn't able to get the information I needed using the `transformer` so I resorted to explore the whole tree manually, from top to bottom. 

This can be optimized to basically  `O(n) ` by running the transformer only on the top level block (and with a few other modifications). I am not aware of any way to do this however.

Reply:
This point needs to be addressed one way or another before merging.

Maybe a solution to prevent the quadratic behavior would be to cut the inner traversal to not got into nested If & Block (or accept a bounded level of nesting).